arch/riscv/src/lib.rs

```diff
@@ -364,8 +370,208 @@ pub extern "C" fn _start_trap() {
     }
 }
 
+/// This is the trap handler function. This code is called on all traps,
+/// including interrupts, exceptions, and system calls from applications.
+///
+/// Tock uses only the single trap handler, and does not use any vectored
+/// interrupts or other exception handling. The trap handler has to determine
+/// why the trap handler was called, and respond accordingly. Generally, there
+/// are two reasons the trap handler gets called: an interrupt occurred or an
+/// application called a syscall.
+///
+/// In the case of an interrupt while the kernel was executing we only need to
+/// save the kernel registers and then run whatever interrupt handling code we
+/// need to. If the trap happens while and application was executing, we have to
+/// save the application state and then resume the `switch_to()` function to
+/// correctly return back to the kernel.
+#[cfg(all(target_arch = "riscv64", target_os = "none"))]
+#[link_section = ".riscv.trap"]
+#[export_name = "_start_trap"]
+#[naked]
+pub extern "C" fn _start_trap() {
```

Comment:
I don't fully understand why this function, for example, is copied.

Reply:
Ah ok, sorry I looked at this on my phone initially. Those copied blocks that use essentially #ifdefs to include different code based on the architecture target worry me.

One concern of course is them becoming out-of-sync (or perhaps they might need to differ in the future). But my primary concern is that it hacks around Rust's dependency system: two boards can have the exact same code in their board crate (and cargo.toml files), but then target different architectures (in a Makefile) and compile very different code blocks. I would expect that at least a different function is called or a different dependency was included to get different code to compile.

I think this PR is proof that a single riscv crate can exist. But I'm not sure it should in Tock.